### PR TITLE
Sample for data pre-processing using azure data factory and azure batch

### PR DESCRIPTION
**Pull request by sapinderpalsingh**
# Type of PR

- Documentation changes
- Code changes

## Purpose
This PR adds a new datafactory sample which demonstrates data pre-processing using container workloads on azure batch.
- ...

## Author pre-publish checklist

- [ ] Added test to prove my fix is effective or new feature works
- [ ] No PII in logs
- [x] Made corresponding changes to the documentation

## Validation steps

- ...

## Issues Closed or Referenced

- Closes #issue_number

- References #issue_number

### CODE REVIEW

File: single_tech_samples/datafactory/README.md [hunk 1]
```diff
@@ -3,3 +3,4 @@
 ## Samples
 
 1. [CICD and automated integration tests for Azure Data Factory](sample1_cicd/README.md) - this demonstrates how to do continuous integration and continuous delivery with Azure Data Factory, along with automated integration tests.
+1. [Running container workloads from Azure Data Factory on Azure Batch for data pre-processing](sample2_data_pre_processing_with_azure_batch/README.md) - this demonstrates how to do continuous integration and continuous delivery with Azure Data Factory, along with automated integration tests.
```
Comment (devlace) [inline single_tech_samples/datafactory/README.md:6]: The description is not right. Copy paste? :)

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    - [2.1.1 Software Prerequisites](#211-software-prerequisites)
+  - [2.2. Setup and deployment](#22-setup-and-deployment)  
+  - [2.3. Deployment validation and Execution](#23-deployment-validation-and-execution)
+  - [2.4. Clean-up](#24-clean-up)
+
+## 1. Solution Overview
+
+This solution demonstrates how we can perform data pre-processing by running Azure Batch container workloads from Azure Data Factory.
+
+This sample will focus on provisioning an Azure Batch account, Azure Data Factory and other required resources where you can run the ADF pipeline to see how the data pre-processing can be done on huge volume of data in an efficient and scalable manner.
```
Comment (devlace) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/README.md:21]: I think you should also highlight _why_ this particular sample architecture of setting up Azure batch with ADF is super useful to get around the limitations of ADF and Azure Batch orchestration, such as the slow spin-up times (thus need for containers), and the lack of ADF support for Azure Batch containers.

**Comment by promisinganuj** [comment]
Hi @sapinderpalsingh, can you please move this as sample 3? There is already a sample 2 which is being merged for "data factory". https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/569

**Review by promisinganuj — CHANGES_REQUESTED** [review]

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/terraform/modules/azureBatch/main.tf [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  tags = var.tags
+}
+
+resource "azurerm_public_ip" "orch_pool_ip" {
+  name                = "${var.orch_pool_name}-ip"
+  resource_group_name = var.resource_group_name
+  location            = var.location
+  allocation_method   = "Static"
+  sku                 = var.ip_sku
+  domain_name_label   = "${var.resource_group_name}-${azurerm_batch_account.batch_account.name}-${var.orch_pool_name}-ip"
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/terraform/modules/azureBatch/main.tf:21]: Can we rename it to something small? I used a slightly bigger resource group name and got the following error:

```
module.azure_batch.azurerm_batch_pool.exec_pool: Creation complete after 1m35s [id=/subscriptions/abc/resourceGroups/rg-databricks-sample/providers/Microsoft.Batch/batchAccounts/batchaccountdevmqxqqxej/pools/executionpool]
╷
│ Error: domain_name_label must contain only lowercase alphanumeric characters, numbers and hyphens. It must start with a letter, end only with a number or letter and not exceed 63 characters in length
│ 
│   with module.azure_batch.azurerm_public_ip.orch_pool_ip,
│   on modules/azureBatch/main.tf line 21, in resource "azurerm_public_ip" "orch_pool_ip":
│   21:   domain_name_label   = "${var.resource_group_name}-${azurerm_batch_account.batch_account.name}-${var.orch_pool_name}-ip"
```
```
"domain_name_label": "rg-databricks-sample-batchaccountdevmqxqqxej-executionpool-ip"
```

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/adf/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Deploy sample ADF pipeline and required linked services
+
+1. `cd modern-data-warehouse-dataops/single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/adf`
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/adf/README.md:3]: cd modern-data-warehouse-dataops/single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/**deploy**/adf

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/adf/deploy-pipeline.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+AZ_RESOURCE_GROUP="<YOUR-RESOURCEGROUP-NAME>"
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/adf/deploy-pipeline.sh:1]: add "#!/bin/sh"

Comment (promisinganuj) [reply single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/deploy/adf/deploy-pipeline.sh:1]: When cloning this repo on MAC, this script doesn't have execute privilege, please add that.
```
chmod +x deploy-pipeline.sh
```

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    ```
+
+### Testing your image locally
+
+Run the following command, which mounts src/sample-processor/data to your container.
+
+`Make sure to update <YOURPATH> with your clonned directory path`
+
+```shell
+docker run --rm --mount type=bind,source=<YOURPATH>/src/sample-processor/data,target=/data sample-processor:latest bash -c "source /opt/ros/noetic/setup.bash&&python3 /code/app.py /data/raw/sample.bag /data/extracted && rosbag info '/data/raw/sample-data.bag' > /data/extracted/rosbagInfo.txt"
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md:26]: This didn't work for me.

- There is no "/code" folder for /code/app.py".
- Even after fixing it, somehow it didn't work.

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    ```shell
+    RESOURCE_GROUP_NAME="<YOUR-RESOURCE-GROUP-NAME>"
+    STORAGE_ACCOUNT_NAME="<YOUR-ADLS-STORAGE-ACCOUNT>"
+
+    #Add your client ip to access storage account.
+    IP_ADDRESS=$(curl -s checkip.dyndns.org | sed -e 's/.*Current IP Address: //' -e 's/<.*$//')
+    az storage account network-rule add -g $RESOURCE_GROUP_NAME --account-name $STORAGE_ACCOUNT_NAME --ip-address $IP_ADDRESS
+
+    #Upload sample file
+    az storage blob upload -f "data/raw/sample-data.bag" -c data/raw --account-name "$STORAGE_ACCOUNT_NAME"
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md:69]:
```
(base) sample-processor (main) >> az storage blob upload -f "data/raw/sample-data.bag" -c data/raw --account-name "$STORAGE_ACCOUNT_NAME"

There are no credentials provided in your command and environment, we will query for account key for your storage account.
It is recommended to provide --connection-string, --account-key or --sas-token in your command as credentials.

You also can add `--auth-mode login` in your command to use Azure Active Directory (Azure AD) for authorization if your login account is assigned required RBAC roles.
For more information about RBAC roles in storage, visit https://docs.microsoft.com/azure/storage/common/storage-auth-aad-rbac-cli.

In addition, setting the corresponding environment variables can avoid inputting credentials in your command. Please use --help to get more information about environment variable usage.
The specified container does not exist.
RequestId:692ab1d5-f01e-0063-2dfe-70f08d000000
Time:2023-04-17T07:33:39.1255557Z
ErrorCode:ContainerNotFound
```

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+`Make sure to update <YOURPATH> with your clonned directory path`
+
+```shell
+docker run --rm --mount type=bind,source=<YOURPATH>/src/sample-processor/data,target=/data sample-processor:latest bash -c "source /opt/ros/noetic/setup.bash&&python3 /code/app.py /data/raw/sample.bag /data/extracted && rosbag info '/data/raw/sample-data.bag' > /data/extracted/rosbagInfo.txt"
+```
+
+Once this command runs successfully, you will see the output in the `src/sample-processor/data/extracted` folder.
+
+### Pushing the docker image to Azure Container Regsitry(ACR)
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md:31]: Can we just put it in a single script and provide the details here? Easy to run.

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md (hunk 6 above)
Comment (sapinderpalsingh) [reply single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md]: Yes, it was missing the parameter names -i and -o

File: single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md (hunk 8 above)
Comment (sapinderpalsingh) [reply single_tech_samples/datafactory/sample2_data_pre_processing_with_azure_batch/src/sample-processor/README.md]: Done

File: single_tech_samples/datafactory/sample3_data_pre_processing_with_azure_batch/README.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+- **Azure Container Registry(ACR):**  In this sample execution pool is configured with a azure container registry to pull container images, but we can have any container registry configured.
+
+- **Managed Identity:** This is a user assigned managed identity which will be assigned to both of the batch pools and batch pool will use it to access any other required resources like container registry, keyvault, storage account etc.
+
+### 1.4. Technologies used
+
+The following technologies are used to build this sample:
+
+- [Azure Data Factory](https://azure.microsoft.com/en-in/products/data-factory/)
```
Comment (promisinganuj) [inline single_tech_samples/datafactory/sample3_data_pre_processing_with_azure_batch/README.md:103]: Remove locale information from URLs:
https://azure.microsoft.com/en-in/products/data-factory/
=>
https://azure.microsoft.com/products/data-factory/

**Review by promisinganuj — CHANGES_REQUESTED** [review]
Suggesting some minor text changes.

**Pull request merged**